kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -0,0 +1,156 @@
+        this.protocol = protocol;
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Row getState(TableClient tableClient) {
+        return TransactionStateRow.of(metadata, dataPath.toString());
+    }
+
+    @Override
+    public List<String> getPartitionColumns(TableClient tableClient) {
+        return newPartitionColumns.orElseGet(() ->
+                VectorUtils.toJavaList(metadata.getPartitionColumns()));
+    }
+
+    @Override
+    public StructType getSchema(TableClient tableClient) {
+        return newSchema.orElseGet(() -> readSnapshot.getSchema(tableClient));
+    }
+
+    @Override
+    public TransactionCommitResult commit(
+            TableClient tableClient,
+            CloseableIterable<Row> dataActions) throws ConcurrentWriteException {
+        checkState(
+                !closed,
+                "Transaction is already attempted to commit. Create a new transaction.");
+        List<Row> metadataActions = new ArrayList<>();
+        metadataActions.add(createCommitInfoSingleAction(generateCommitAction()));
+        metadataActions.add(createMetadataSingleAction(metadata.toRow()));
+        metadataActions.add(createProtocolSingleAction(protocol.toRow()));
+
+
+        try (CloseableIterator<Row> stageDataIter = dataActions.iterator()) {
+            // Create a new CloseableIterator that will return the metadata actions followed by the
+            // data actions.
+            CloseableIterator<Row> stagedDataWithCommitInfo =
+                    toCloseableIterator(metadataActions.iterator()).combine(stageDataIter);
+
+            try {
+                long readVersion = readSnapshot.getVersion(tableClient);
+                if (readVersion == -1) {
+                    // New table, create a delta log directory
+                    if (!tableClient.getFileSystemClient().mkdirs(logPath.toString())) {
+                        throw new RuntimeException(
+                                "Failed to create delta log directory: " + logPath);
+                    }
+                }
+
+                long newVersion = readVersion + 1;
+                // Write the staged data to a delta file
+                tableClient.getJsonHandler().writeJsonFileAtomically(
+                        FileNames.deltaFile(logPath, newVersion),
+                        stagedDataWithCommitInfo,
+                        false /* overwrite */);
+
+                return new TransactionCommitResult(newVersion, isReadyForCheckpoint(newVersion));
+            } catch (FileAlreadyExistsException e) {
+                // TODO: Resolve conflicts and retry commit
+                throw new ConcurrentWriteException();
+            }
+        } catch (IOException ioe) {
+            throw new RuntimeException(ioe);
+        } finally {
+            closed = true;
+        }
+    }
+
+    private Row generateCommitAction() {
+        return new CommitInfo(
+                System.currentTimeMillis(),
+                engineInfo,
```

Comment:
I think this is where we want to inject the kernel engine information.

Reply:
Added the code.